Allow omitting the target triple for `swift sdk configure`

marcprux opened the pull request:
You can now run `swift sdk configure --sdk-root-path /some/sdk/root sdk-id` without needing to specify the target triple

### Motivation:

The Android SDK (https://github.com/swiftlang/swift/issues/80788#issuecomment-2840356225) needs to have its `sdkRootPath` configured to point to an external NDK sysroot. But it contains many different target triples, like `armv7-unknown-linux-androideabi33` and `x86_64-unknown-linux-android29`, all of which need to be configured to point to the same sysroot. Following on to https://github.com/swiftlang/swift-package-manager/pull/8687 and https://github.com/swiftlang/swift-evolution/pull/2888, this PR enables running a single command to configure the entire SDK, like so:

```
swift run swift-sdk configure --sdk-root-path ~/Library/Android/sdk/ndk/27.0.12077973/toolchains/llvm/prebuilt/darwin-x86_64/sysroot swift-6.1-RELEASE-android-24-0.1
```

This will result in configurations being created for each of the target triples for the SDK ID.

<details>

```
% swift run swift-sdk configure --sdk-root-path ~/Library/Android/sdk/ndk/27.0.12077973/toolchains/llvm/prebuilt/darwin-x86_64/sysroot swift-6.1-RELEASE-android-24-0.1
info: These properties of Swift SDK `swift-6.1-RELEASE-android-24-0.1` for target triple `aarch64-unknown-linux-android25` were successfully updated: sdkRootPath.
info: These properties of Swift SDK `swift-6.1-RELEASE-android-24-0.1` for target triple `armv7-unknown-linux-androideabi33` were successfully updated: sdkRootPath.
info: These properties of Swift SDK `swift-6.1-RELEASE-android-24-0.1` for target triple `aarch64-unknown-linux-android26` were successfully updated: sdkRootPath.
info: These properties of Swift SDK `swift-6.1-RELEASE-android-24-0.1` for target triple `armv7-unknown-linux-androideabi35` were successfully updated: sdkRootPath.
info: These properties of Swift SDK `swift-6.1-RELEASE-android-24-0.1` for target triple `aarch64-unknown-linux-android31` were successfully updated: sdkRootPath.
info: These properties of Swift SDK `swift-6.1-RELEASE-android-24-0.1` for target triple `armv7-unknown-linux-androideabi29` were successfully updated: sdkRootPath.
info: These properties of Swift SDK `swift-6.1-RELEASE-android-24-0.1` for target triple `aarch64-unknown-linux-android32` were successfully updated: sdkRootPath.
info: These properties of Swift SDK `swift-6.1-RELEASE-android-24-0.1` for target triple `aarch64-unknown-linux-android34` were successfully updated: sdkRootPath.
info: These properties of Swift SDK `swift-6.1-RELEASE-android-24-0.1` for target triple `x86_64-unknown-linux-android27` were successfully updated: sdkRootPath.
info: These properties of Swift SDK `swift-6.1-RELEASE-android-24-0.1` for target triple `armv7-unknown-linux-androideabi25` were successfully updated: sdkRootPath.
info: These properties of Swift SDK `swift-6.1-RELEASE-android-24-0.1` for target triple `armv7-unknown-linux-androideabi26` were successfully updated: sdkRootPath.
info: These properties of Swift SDK `swift-6.1-RELEASE-android-24-0.1` for target triple `armv7-unknown-linux-androideabi24` were successfully updated: sdkRootPath.
info: These properties of Swift SDK `swift-6.1-RELEASE-android-24-0.1` for target triple `x86_64-unknown-linux-android33` were successfully updated: sdkRootPath.
info: These properties of Swift SDK `swift-6.1-RELEASE-android-24-0.1` for target triple `aarch64-unknown-linux-android35` were successfully updated: sdkRootPath.
info: These properties of Swift SDK `swift-6.1-RELEASE-android-24-0.1` for target triple `x86_64-unknown-linux-android29` were successfully updated: sdkRootPath.
info: These properties of Swift SDK `swift-6.1-RELEASE-android-24-0.1` for target triple `x86_64-unknown-linux-android32` were successfully updated: sdkRootPath.
info: These properties of Swift SDK `swift-6.1-RELEASE-android-24-0.1` for target triple `x86_64-unknown-linux-android25` were successfully updated: sdkRootPath.
info: These properties of Swift SDK `swift-6.1-RELEASE-android-24-0.1` for target triple `aarch64-unknown-linux-android24` were successfully updated: sdkRootPath.
info: These properties of Swift SDK `swift-6.1-RELEASE-android-24-0.1` for target triple `armv7-unknown-linux-androideabi30` were successfully updated: sdkRootPath.
info: These properties of Swift SDK `swift-6.1-RELEASE-android-24-0.1` for target triple `x86_64-unknown-linux-android26` were successfully updated: sdkRootPath.
info: These properties of Swift SDK `swift-6.1-RELEASE-android-24-0.1` for target triple `aarch64-unknown-linux-android33` were successfully updated: sdkRootPath.
info: These properties of Swift SDK `swift-6.1-RELEASE-android-24-0.1` for target triple `armv7-unknown-linux-androideabi32` were successfully updated: sdkRootPath.
info: These properties of Swift SDK `swift-6.1-RELEASE-android-24-0.1` for target triple `x86_64-unknown-linux-android35` were successfully updated: sdkRootPath.
info: These properties of Swift SDK `swift-6.1-RELEASE-android-24-0.1` for target triple `x86_64-unknown-linux-android34` were successfully updated: sdkRootPath.
info: These properties of Swift SDK `swift-6.1-RELEASE-android-24-0.1` for target triple `x86_64-unknown-linux-android31` were successfully updated: sdkRootPath.
info: These properties of Swift SDK `swift-6.1-RELEASE-android-24-0.1` for target triple `aarch64-unknown-linux-android28` were successfully updated: sdkRootPath.
info: These properties of Swift SDK `swift-6.1-RELEASE-android-24-0.1` for target triple `aarch64-unknown-linux-android27` were successfully updated: sdkRootPath.
info: These properties of Swift SDK `swift-6.1-RELEASE-android-24-0.1` for target triple `x86_64-unknown-linux-android24` were successfully updated: sdkRootPath.
info: These properties of Swift SDK `swift-6.1-RELEASE-android-24-0.1` for target triple `x86_64-unknown-linux-android30` were successfully updated: sdkRootPath.
info: These properties of Swift SDK `swift-6.1-RELEASE-android-24-0.1` for target triple `armv7-unknown-linux-androideabi28` were successfully updated: sdkRootPath.
info: These properties of Swift SDK `swift-6.1-RELEASE-android-24-0.1` for target triple `x86_64-unknown-linux-android28` were successfully updated: sdkRootPath.
info: These properties of Swift SDK `swift-6.1-RELEASE-android-24-0.1` for target triple `armv7-unknown-linux-androideabi27` were successfully updated: sdkRootPath.
info: These properties of Swift SDK `swift-6.1-RELEASE-android-24-0.1` for target triple `aarch64-unknown-linux-android30` were successfully updated: sdkRootPath.
info: These properties of Swift SDK `swift-6.1-RELEASE-android-24-0.1` for target triple `armv7-unknown-linux-androideabi34` were successfully updated: sdkRootPath.
info: These properties of Swift SDK `swift-6.1-RELEASE-android-24-0.1` for target triple `armv7-unknown-linux-androideabi31` were successfully updated: sdkRootPath.
info: These properties of Swift SDK `swift-6.1-RELEASE-android-24-0.1` for target triple `aarch64-unknown-linux-android29` were successfully updated: sdkRootPath.

% ls ~/Library/org.swift.swiftpm/swift-sdks/configuration/
swift-6.1-RELEASE-android-24-0.1_aarch64-unknown-linux-android24.json
swift-6.1-RELEASE-android-24-0.1_aarch64-unknown-linux-android25.json
swift-6.1-RELEASE-android-24-0.1_aarch64-unknown-linux-android26.json
swift-6.1-RELEASE-android-24-0.1_aarch64-unknown-linux-android27.json
swift-6.1-RELEASE-android-24-0.1_aarch64-unknown-linux-android28.json
swift-6.1-RELEASE-android-24-0.1_aarch64-unknown-linux-android29.json
swift-6.1-RELEASE-android-24-0.1_aarch64-unknown-linux-android30.json
swift-6.1-RELEASE-android-24-0.1_aarch64-unknown-linux-android31.json
swift-6.1-RELEASE-android-24-0.1_aarch64-unknown-linux-android32.json
swift-6.1-RELEASE-android-24-0.1_aarch64-unknown-linux-android33.json
swift-6.1-RELEASE-android-24-0.1_aarch64-unknown-linux-android34.json
swift-6.1-RELEASE-android-24-0.1_aarch64-unknown-linux-android35.json
swift-6.1-RELEASE-android-24-0.1_armv7-unknown-linux-androideabi24.json
swift-6.1-RELEASE-android-24-0.1_armv7-unknown-linux-androideabi25.json
swift-6.1-RELEASE-android-24-0.1_armv7-unknown-linux-androideabi26.json
swift-6.1-RELEASE-android-24-0.1_armv7-unknown-linux-androideabi27.json
swift-6.1-RELEASE-android-24-0.1_armv7-unknown-linux-androideabi28.json
swift-6.1-RELEASE-android-24-0.1_armv7-unknown-linux-androideabi29.json
swift-6.1-RELEASE-android-24-0.1_armv7-unknown-linux-androideabi30.json
swift-6.1-RELEASE-android-24-0.1_armv7-unknown-linux-androideabi31.json
swift-6.1-RELEASE-android-24-0.1_armv7-unknown-linux-androideabi32.json
swift-6.1-RELEASE-android-24-0.1_armv7-unknown-linux-androideabi33.json
swift-6.1-RELEASE-android-24-0.1_armv7-unknown-linux-androideabi34.json
swift-6.1-RELEASE-android-24-0.1_armv7-unknown-linux-androideabi35.json
swift-6.1-RELEASE-android-24-0.1_x86_64-unknown-linux-android24.json
swift-6.1-RELEASE-android-24-0.1_x86_64-unknown-linux-android25.json
swift-6.1-RELEASE-android-24-0.1_x86_64-unknown-linux-android26.json
swift-6.1-RELEASE-android-24-0.1_x86_64-unknown-linux-android27.json
swift-6.1-RELEASE-android-24-0.1_x86_64-unknown-linux-android28.json
swift-6.1-RELEASE-android-24-0.1_x86_64-unknown-linux-android29.json
swift-6.1-RELEASE-android-24-0.1_x86_64-unknown-linux-android30.json
swift-6.1-RELEASE-android-24-0.1_x86_64-unknown-linux-android31.json
swift-6.1-RELEASE-android-24-0.1_x86_64-unknown-linux-android32.json
swift-6.1-RELEASE-android-24-0.1_x86_64-unknown-linux-android33.json
swift-6.1-RELEASE-android-24-0.1_x86_64-unknown-linux-android34.json
swift-6.1-RELEASE-android-24-0.1_x86_64-unknown-linux-android35.json

% cat ~/Library/org.swift.swiftpm/swift-sdks/configuration/swift-6.1-RELEASE-android-24-0.1_x86_64-unknown-linux-android31.json
{
  "sdkRootPath" : "/Users/marc/Library/Android/sdk/ndk/27.0.12077973/toolchains/llvm/prebuilt/darwin-x86_64/sysroot",
  "swiftResourcesPath" : "/Users/marc/Library/org.swift.swiftpm/swift-sdks/swift-6.1-RELEASE-android-24-0.1.artifactbundle/swift-6.1-release-android-24-sdk/android-27c-sysroot/usr/lib/swift",
  "swiftStaticResourcesPath" : "/Users/marc/Library/org.swift.swiftpm/swift-sdks/swift-6.1-RELEASE-android-24-0.1.artifactbundle/swift-6.1-release-android-24-sdk/android-27c-sysroot/usr/lib/swift_static-x86_64",
  "toolsetPaths" : [
    "/Users/marc/Library/org.swift.swiftpm/swift-sdks/swift-6.1-RELEASE-android-24-0.1.artifactbundle/swift-6.1-release-android-24-sdk/swift-toolset.json"
  ]
}
```
</details>

In addition, this fixes https://github.com/swiftlang/swift-package-manager/issues/8584, where `swift sdk configure` with a specified target triple didn't work at all, and instead would configure some random target triple for the specified SDK.

### Modifications:

Make the `targetTriple` flag for `ConfigureSwiftSDK` optional and, when nil, apply the command to all triples included in the specified SDK.

### Result:

Fixes https://github.com/swiftlang/swift-package-manager/issues/8584 and enables configuring multiple target triples with a single `swift sdk configure` command.

marcprux commented:
CC @finagolfin, @etcwilde, @MaxDesiatov

finagolfin commented:
@swift-ci test self hosted

marcprux commented:
> I think the reviewers will ask for a test

There's no existing test support for the `sdk configure` command. I could probably implement something, but it would be a fairly major refactor, since much of the `ConfigureSwiftSDK.run()` logic would need to be relocated to somewhere more testable, like `SwiftSDKConfigurationStore`.

I'd be willing to undertake it if the reviewers want/need it.

MaxDesiatov commented:
> > I think the reviewers will ask for a test
> 
> There's no existing test support for the `sdk configure` command. I could probably implement something, but it would be a fairly major refactor, since much of the `ConfigureSwiftSDK.run()` logic would need to be relocated to somewhere more testable, like `SwiftSDKConfigurationStore`.
> 
> I'd be willing to undertake it if the reviewers want/need it.

Yes please!

marcprux commented:
OK, I'll work on implementing test support.

Are there any other common use cases for `swift sdk configure`, either existing or envisioned? I'd like to make sure my tests cover as many plausible scenarios as possible, and not just the Android SDK sysroot setup.

MaxDesiatov commented:
> OK, I'll work on implementing test support.
> 
> Are there any other common use cases for `swift sdk configure`, either existing or envisioned? I'd like to make sure my tests cover as many plausible scenarios as possible, and not just the Android SDK sysroot setup.

Not at the moment, I've designed it almost exclusively to support out-of-tree sysroots like NDK 😅

marcprux commented:
I've added a `SwiftSDKBundleTests.testConfigureSDKRootPath` test, which is supported by moving the SDK configuration logic from the `ConfigureSwiftSDK` command to the `SwiftSDKConfigurationStore`. In doing so, I also centralized the config properties by making `SwiftSDK.PathsConfiguration` generic, so `String` and `AbsolutePath` variants will share the same structure; should new configuration properties ever need to be added in the future, it will be simpler to have them in this one place.

Lastly, I cleaned up some of the logic for the `PathsConfiguration` initializers which can accept an optional `swiftSDKDirectory` base directory and condensed it by eliminating the branch on nil to instead use the `AbsolutePath` initializer extension that accepts an optional `relativeTo` argument (which I had to relocate there anyway).

finagolfin commented:
@swift-ci test self hosted

MaxDesiatov reviewed (commented):
Thanks! Other than formatting nits (see [`CONTRIBUTING.md` "Creating PRs" section, p. 7](https://github.com/swiftlang/swift-package-manager/blob/main/CONTRIBUTING.md#creating-pull-requests)) and unnecessary use of `public`, I think that new and existing functions and types deserve more doc comments while we're at it.

marcprux commented on Sources/SwiftSDKCommand/ConfigureSwiftSDK.swift:
```diff
@@ -132,101 +132,21 @@ struct ConfigureSwiftSDK: AsyncParsableCommand {
+                showConfiguration: shouldShowConfiguration,
+                resetConfiguration: shouldReset,
+                config: config
+            ) == false {
```
OK, fixed.

MaxDesiatov commented on Sources/PackageModel/SwiftSDKs/SwiftSDKConfigurationStore.swift:
```diff
@@ -141,4 +157,92 @@ public final class SwiftSDKConfigurationStore {
+        if let targetTriple = targetTriple {
+            targetTriples = try [Triple(targetTriple)]
+        } else {
+            // when target-triple is unspecified, configure every triple for the SDK
```
comment formatting nit
```suggestion
            // when `targetTriple` is unspecified, configure every triple for the SDK
```

MaxDesiatov commented on Sources/PackageModel/SwiftSDKs/SwiftSDKConfigurationStore.swift:
```diff
@@ -158,7 +158,15 @@ public final class SwiftSDKConfigurationStore {
     }
 
-    public func configure(
+    /// Configures the specified SDK and identified target triple with the configuration parameter.
+    /// - Parameters:
+    ///   - sdkID: ID of the Swift SDK to operate on.
+    ///   - tripleString: run-time triple for which the properties should be configured, or nil to configure all triples for the SDK
```
"Swift SDK" is a SwiftPM concept, while "SDK" is an Xcode and partially a Clang concept, this is also reflected in the difference between `--sdk` and `--swift-sdk` CLI options. We shouldn't mix these two up and should always refer to Swift SDKs as "Swift SDK".

```suggestion
    /// Configures the specified Swift SDK and identified target triple with the configuration parameter.
    /// - Parameters:
    ///   - sdkID: ID of the Swift SDK to operate on.
    ///   - tripleString: run-time triple for which the properties should be configured, or nil to configure all triples for the Swift SDK
```

MaxDesiatov commented on Tests/PackageModelTests/SwiftSDKBundleTests.swift:
```diff
@@ -551,4 +551,93 @@ final class SwiftSDKBundleTests: XCTestCase {
+            let (config, fileSystem) = try await createConfigurationStore()
+            var args = SwiftSDK.PathsConfiguration<String>()
+            args.sdkRootPath = "/some/sdk/root/path"
+            let configSuccess = try config.configure(sdkID: testArtifactID, targetTriple: targetTriple.tripleString, showConfiguration: false, resetConfiguration: false, config: args)
```
I suspect this line is still over the limit in terms of length. Would you mind applying SwiftFormat on this file per `CONTRIBUTING.md`?

MaxDesiatov replied:
SwiftPM doesn't use swift-format, but SwiftFormat, could that be the reason?

marcprux replied:
Doh! Of course. 🙄

MaxDesiatov commented:
@swift-ci test

MaxDesiatov commented:
@swift-ci test windows

marcprux commented:
> @swift-ci test windows

I see that the Windows CI failed, but the link returns 404. Maybe the log expired? Can you run it again and I can take a look?

MaxDesiatov commented:
@swift-ci test self hosted windows

marcprux commented:
Windows error was:

```swift
11:51:14  C:\source\swiftpm\Tests\PackageModelTests\SwiftSDKBundleTests.swift:635: error: SwiftSDKBundleTests.testConfigureSDKRootPath : XCTAssertEqual failed: ("Optional("/some/sdk/root/path")") is not equal to ("Optional("\\some\\sdk\\root\\path")") - 
```

I fixed it in b155f05. Can I get another CI run?

finagolfin commented:
@swift-ci test self hosted windows

marcprux commented:
CI passed. Is there anything else preventing this PR from getting merged?

MaxDesiatov commented:
It needs a review from someone on the SwiftPM team.

MaxDesiatov commented:
@swift-ci test windows

MaxDesiatov commented:
@swift-ci test windows

MaxDesiatov commented:
@marcprux would you have a moment to resolve conflicts? At least GitHub says there are some without showing specific files.

marcprux commented:
@MaxDesiatov Resolved the conflict (in [Sources/PackageManagerDocs/Documentation.docc/SDK/SDKConfigure.md](https://github.com/swiftlang/swift-package-manager/pull/8856/files#diff-f1fb0a9b7ef394bdf316b1fbecd83f6052fc30ace6d8ffffe86870c272502941))

MaxDesiatov commented:
@swift-ci test

jakepetroules commented:
Change looks good, but there is a merge conflict to resolve.

finagolfin commented:
@swift-ci test
